Summer 2017

AdamBJ opened the pull request:
This PR improves upon the README, fixes grammatical errors in comments, fixes a bug in UmpleGradlePlugin.groovy that was causing an `IllegalArugmentException` to be thrown when `relativize` was called, and includes bug fixes relating to the integrationTest task (we should now once again be able to make the integration tests a normal part of the build process).

This branch isn't quite ready to be merged yet, have a look at line 27 of TestProject's `build.gradle` file (`src/integrationTest/resources/TestProject/build.gradle`). For some reason gradle thinks that the outputs of the `compileUmple` task are up-to-date all the time, even if the output files have been deleted. I was only able to get it working by apply the band aid you see starting at line 27. I was looking [here](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_inputs_outputs) to try and figure out what the problem was, but no luck yet. I think the solution may be annotating `UmpleGenerateTask` differently?

AdamBJ committed:
Refactoring, bug fixes

Improve README, fix grammatical errors in comments, fixed bug in UmpleGradlePlugin.groovy that was causing an IllegalArugmentException to be thrown when relativize was called, bug fixes relating to the integrationTest task.

AdamBJ committed:
Further README improvements

AdamBJ commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -21,18 +21,23 @@ apply plugin: 'java'
-  
+
+// TODO this should not be neccesary. Find out why compileUmple is returning "UP-TO-DATE" when the output files have been deleted. 
+compileUmple {
```
We should get rid of this work around before merging this PR.

Nava2 replied:
This TODO should be fixed by using [IncrementalTaskInputs](https://docs.gradle.org/3.1/javadoc/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html). You can see how I used them here: https://github.com/assertj/assertj-generator-gradle-plugin/blob/9f0ec19e910a3d7a4add5de461481454688063d8/src/main/groovy/org/assertj/generator/gradle/tasks/AssertJGenerationTask.groovy#L67

AdamBJ replied:
I'm not sure if IncrementalTaskInputs are the answer. The problem isn't that gradle is mistaking a modified/removed input file for an unmodified input file, the problem is that gradle isn't checking whether generated/output files (i.e. java files) exist when it does it's "up to date" check for the `compileUmple` task. We really just need gradle to check to see if the output files have been removed, and if they have, to rerun the generation task.

Nava2 replied:
That's what IncrementalTaskInputs does. If you register all of the produced files with ITI, it will remove/update them when required.

AdamBJ committed:
Java compilation bug fix

- Made the `compileUmple` task be finalized by the `compileJava` task. This allows the user to run `compileUmple` to generate java files and compile them (so long as the `compileGenerated` flag is set). Before, `compileJava` depended on `compileUmple`, which meant that `compileJava` needed to be executed in order for generation and compilation to occur. This changes also makes the integrationTest task a bit simplier.

AdamBJ commented:
I made another [commit ](https://github.com/umple/umple.gradle/pull/22/commits/2f843f7062d167d803508488b4857c8b52b52b40)today, small but important bug fix for the way generation followed by compilation works.

AdamBJ committed:
Adding incremental build support to the UmpleGenerateTask

AdamBJ commented on src/main/groovy/cruise/umple/tasks/UmpleGenerateTask.groovy:
```diff
@@ -4,20 +4,40 @@ import cruise.umple.UmpleConsoleConfig
 /**
  * Created by kevin on 15/03/2017.
  */
 class UmpleGenerateTask extends SourceTask {
```
@Nava2 The incremental build stuff appears to be working now. I think this branch can be merged into ss-config-n2, now?

Nava2 replied:
I don't think that this is properly implemented. I think Gradle needs to know _all_ files that are generated. We either need to keep track of the files created or figure out a way to have Umple tell us what files it will change/create. The intention of the incremental updates is that gradle can choose whether or not to generate new files and also what files to delete.

AdamBJ replied:
> I think Gradle needs to know all files that are generated

Doesn't the `outputDirs` member variable tell it that?

AdamBJ replied:
As far as I can tell, other than the "nice to have" refactoring of `integrationTest`, this is the only real issue that needs to be resolved before merging this branch. What do you think of my earlier comment? I agree the current implementation looks a bit funky, but by setting up the `outputDirs` member variable like that, doesn't Gradle know where all the generated files will be?

Nava2 replied:
No because it needs to know the files directly. Thats what the incremental input files parameter does.

AdamBJ replied:
Looking at the example here: https://docs.gradle.org/3.1/javadoc/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html

What if we have an `inputDirs` member variable that contains the locations of each of the `master` files in addition to the `outputDirs` member variable that we have now? Then we'd basically just copy the `if (!inputs.incremental)` and `inputs.outOfDate...` stuff from the example... and hopefully we'd be on our way.

Nava2 replied:
Yes, somewhat. That assumes that we want to always change all files. I don't think we have another choice. The way the umple compiler works requires us to regenerate the entire tree no matter what file changes.

So yeah, that should be the best we can do, but not the ideal case. 🙂

AdamBJ replied:
Getting the incremental build stuff working for the integration tests is proving tricky. It works after I rebuild the plugin, but if I don't rebuild it the generation phase gets skipped while the verification step runs anyway (!!?!?!?!), which causes the test to fail. Even better is that the printing to standard out stuff seems to largely be broken for the integration tests, so I'm more or less doing everything blind. It might have something to do with the fact that we're using the[ `test-sets` plugin](https://github.com/unbroken-dome/gradle-testsets-plugin) for `integrationTest`, but I'm really not sure. It looks like I might be able to use `GenerationTests.groovy` as a template if I can figure how to hook everything together.

On an unrelated note, I'll be starting my final semester on Tuesday. I'm not sure how much time I'll have to contribute to this project over the semester, but I'll do my best to pop in every now and again.

The only thing I can think is to try and convert `integrationTest` into a "normal" test (using Project Runner). I'm not sure how to go about doing that, though. How are the current tests (CheckGlobals.groovy etc) hooked into the test task?

AdamBJ commented on src/main/groovy/cruise/umple/tasks/UmpleGenerateTask.groovy:
```diff
@@ -4,20 +4,40 @@ import cruise.umple.UmpleConsoleConfig
+    void setOutputDirs(UmpleOptions newOpts) {
+        this.outputDirs = new ArrayList<File>()
+
+        UmpleOptions opts = newOpts
```
I used the logic in `execute` as the basis for the logic here. I noticed that in `execute` we make a copy of the UmpleOptions (in `setOutputDirs` we do `UmpleOptions opts = newOpts`). Is this copy necessary? I don't think we're changing the `UmpleOption` object.

Nava2 replied:
99% sure that's referential copy so it does _literally_ nothing.

AdamBJ commented:
@Nava2 I've been looking into the Travis failure. The Linux build is working fine, but on Mac there's an assertion error here:

`assertEquals("invalid outputDir",
                Paths.get(testProjectDir.root.toString(), "src/generated/\${language}"),
                Paths.get((String)props.get("umple.outputDir")))
`

I read on the travis site that: "Groovy builds are not available on the OS X environment." (https://docs.travis-ci.com/user/languages/groovy/). So how about we just remove the Mac stuff from Travis?

Nava2 commented:
@AdamBJ I can't get to this till Tuesday I think. Defence + some personal things. Bump this if you don't hear from me by then, though.

Nava2 commented:
Regarding the OSX build, it's probably installable via homebrew, `brew install groovy`.

AdamBJ commented:
Good luck on Monday!

AdamBJ commented:
Bump! Hope the thesis defence went well.

Nava2 commented on Base project/build.gradle:
```diff
@@ -12,6 +12,7 @@ buildscript {
     classpath files('libs/umple-latest.jar')
     //TODO update {CURRENT_VERSION} to the version of the plugin you're using
     classpath "umple.gradle.plugin:UmpleGradlePlugin-{CURRENT_VERSION}" 
+    classpath group: 'com.google.guava', name: 'guava', version: '21.0'
```
Should probably make this 22.

Nava2 commented on libs/placeholder.txt:
```diff
@@ -0,0 +1 @@
+Put the umple jar in this folder. It should be named umple-latest.jar.
```
We could use `git-lfs` if necessary here, but ideally we need to package umple for ivy/maven and ship it that way.

Nava2 commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -8,9 +8,8 @@ buildscript {
   dependencies {
```
Do we still need the `integrationTest` projects as we use the project via the gradle test runner?

AdamBJ replied:
I think we could convert it from the current style to the gradle test runner style we use for everything else (e.g. have the build file included within CompilationAndGeneration.groovy). I'm not 100% sure how it would work, though, since it's one of the more complicated tests (there are subprojects with their own build files, etc). I'd say it would be a "nice to have" feature to have a more consistent testing approach, but that it's probably not worth the effort to switch it over.

AdamBJ replied:
And to be clear, I don't think we use the `integrationTest` project outside of `CompilationAndGeneration.groovy`.

Nava2 replied:
Yeah, I don't think its necessary -- we should be able to use the groovy test runner + temp folders.

Nava2 commented on src/main/groovy/cruise/umple/UmpleGradlePlugin.groovy:
```diff
@@ -164,22 +163,23 @@ class UmpleGradlePlugin implements Plugin<Project> {
-            project.tasks.getByName(sourceSet.compileJavaTaskName).dependsOn umpleGenerate
+            // The user has requested that generated Java files are compiled. Make sure
+            // the compileJava task runs after the compileUmple task
+            umpleGenerate.finalizedBy project.tasks.getByName(sourceSet.compileJavaTaskName)
```
This is over using the umple compile mechanics? If so, I much prefer this -- use Gradle to it's advantage.

AdamBJ replied:
Yeah, when it comes to compiling Java files we don't (and never have as far as I can tell) use the Umple mechanic. This change makes it so the user can set the `compileGenerated` flag and invoke the `compileUmple` task in order to generate and compile. Before the user had to use `compileJava` if they wanted to compile the generated source. It made the `compileGenerated` flag redundant.

Nava2 replied:
I think this should be on by default -- which it is! I can't think of a reason to have it be off by default.

Nava2 reviewed (commented):
Looking for some more feedback on this. But I'd also love to get this branch removed and some work into the master branch instead.

Nava2 commented:
I had this email marked unread because I had to do something.. but I don't know what. Is there something that's needed here other than the comments I've made? I can investigate the mac build.

Nava2 commented:
Comment I didn't make before, change `Base project` folder to something like `examples/` or similar. Just looks strange with all of the other structure.

Nava2 committed:
Rename Base project folder

Nava2 committed:
Adjust build for OSX taken from assertj/assertj-generator-gradle-plugin

Nava2 committed:
Add OSX with updating the brew installations

Nava2 committed:
Change back to matrix style declaration

AdamBJ commented:
> I had this email marked unread because I had to do something.. but I don't know what. Is there something that's needed here other than the comments I've made? I can investigate the mac build.

I think just that business with the incremental build. I'll have another look at that now. Great to see Travis green again!

Nava2 commented:
> I think just that business with the incremental build. I'll have another look at that now. Great to see Travis green again!

Still not green. :( There's an issue with tests on OS X, not sure why. I'm going to look into that when I finish some other work.
